storage/postgreskv/client.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	q := `
+		SELECT metadata
+		FROM pathdata pd
+			RIGHT JOIN
+				unnest($2::BYTEA[]) WITH ORDINALITY pk(request, ord)
+			ON (pd.fullpath = pk.request AND pd.bucket = $1::BYTEA)
+		ORDER BY pk.ord
+	`
+	rows, err := client.pgConn.Query(q, []byte(bucket), pq.ByteaArray(keys.ByteSlices()))
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	values := make([]storage.Value, 0, len(keys))
+	for rows.Next() {
+		var value []byte
+		if err := rows.Scan(&value); err != nil {
+			_ = rows.Close()
+			return nil, errs.Wrap(err)
+		}
+		values = append(values, storage.Value(value))
+	}
+	_ = rows.Close()
+
+	return values, nil
+}
+
+type orderedPostgresIterator struct {
+	client         *Client
+	opts           *storage.IterateOptions
+	bucket         storage.Key
+	delimiter      byte
+	batchSize      int
+	curIndex       int
+	curRows        *sql.Rows
+	lastKeySeen    storage.Key
+	errEncountered error
+}
+
+// Next fills in info for the next item in an ongoing listing.
+func (opi *orderedPostgresIterator) Next(item *storage.ListItem) bool {
+	if !opi.curRows.Next() {
+		_ = opi.curRows.Close()
+		if opi.curIndex < opi.batchSize {
+			return false
+		}
+		if err := opi.curRows.Err(); err != nil {
+			opi.errEncountered = errs.Wrap(err)
+			return false
+		}
+		newRows, err := opi.nextQuery()
+		if err != nil {
+			opi.errEncountered = errs.Wrap(err)
+			return false
+		}
+		opi.curRows = newRows
+		opi.curIndex = 0
+		if !opi.curRows.Next() {
+			_ = opi.curRows.Close()
+			return false
+		}
+	}
+	var k, v []byte
+	err := opi.curRows.Scan(&k, &v)
+	if err != nil {
+		opi.errEncountered = errs.Wrap(err)
+		_ = opi.curRows.Close()
+		return false
+	}
+	item.Key = storage.Key(k)
+	item.Value = storage.Value(v)
+	opi.curIndex++
+	if opi.curIndex == 0 && opi.lastKeySeen.Equal(item.Key) {
+		return opi.Next(item)
+	}
+	if !opi.opts.Recurse && item.Key[len(item.Key)-1] == opi.delimiter && !item.Key.Equal(opi.opts.Prefix) {
+		item.IsPrefix = true
+		// i don't think this makes the most sense, but it's necessary to pass the storage testsuite
+		item.Value = nil
+	} else {
+		item.IsPrefix = false
+	}
+	opi.lastKeySeen = item.Key
+	return true
+}
+
+func (opi *orderedPostgresIterator) nextQuery() (*sql.Rows, error) {
+	start := opi.lastKeySeen
+	if start == nil {
+		start = opi.opts.First
+	}
+	var query string
+	if !opi.opts.Recurse {
+		if opi.opts.Reverse {
+			query = "SELECT p, m FROM list_directory_reverse($1::BYTEA, $2::BYTEA, $3::BYTEA, $4) ld(p, m)"
+		} else {
+			query = "SELECT p, m FROM list_directory($1::BYTEA, $2::BYTEA, $3::BYTEA, $4) ld(p, m)"
+		}
+	} else {
+		startCmp := ">="
+		orderDir := ""
+		if opi.opts.Reverse {
+			startCmp = "<="
+			orderDir = " DESC"
+		}
+		query = fmt.Sprintf(`
+			SELECT fullpath, metadata
+			  FROM pathdata
+			 WHERE bucket = $1::BYTEA
+			   AND ($2::BYTEA = ''::BYTEA OR fullpath >= $2::BYTEA)
+			   AND ($2::BYTEA = ''::BYTEA OR fullpath < bytea_increment($2::BYTEA))
```

Comment:
bytea_increment would still need to be defined in the SQL, though, because it's used in the recursive CTE. i chose to make use of the one in PG here to avoid implementing it twice and running the risk of the two diverging.